probe/awsecs/client.go

```diff
@@ -1,102 +1,212 @@
 }
 
-// cannot fail as it will attempt to deliver partial results, though that may end up being no results
-func (c ecsClient) getServices() map[string]*ecs.Service {
-	results := map[string]*ecs.Service{}
-	lock := sync.Mutex{} // lock mediates access to results
-
-	group := sync.WaitGroup{}
-
-	err := c.client.ListServicesPages(
-		&ecs.ListServicesInput{Cluster: &c.cluster},
-		func(page *ecs.ListServicesOutput, lastPage bool) bool {
-			// describe each page of 10 (the max for one describe command) concurrently
-			group.Add(1)
-			serviceArns := page.ServiceArns
-			go func() {
-				defer group.Done()
-
-				resp, err := c.client.DescribeServices(&ecs.DescribeServicesInput{
-					Cluster:  &c.cluster,
-					Services: serviceArns,
-				})
-				if err != nil {
-					log.Warnf("Error describing some ECS services, ECS service report may be incomplete: %v", err)
-					return
-				}
+func newECSService(service *ecs.Service) EcsService {
+	deploymentIDs := make([]string, len(service.Deployments))
+	for i, deployment := range service.Deployments {
+		deploymentIDs[i] = *deployment.Id
+	}
+	return EcsService{
+		ServiceName:       *service.ServiceName,
+		DeploymentIDs:     deploymentIDs,
+		DesiredCount:      *service.DesiredCount,
+		PendingCount:      *service.PendingCount,
+		RunningCount:      *service.RunningCount,
+		TaskDefinitionARN: *service.TaskDefinition,
+	}
+}
 
-				for _, failure := range resp.Failures {
-					log.Warnf("Failed to describe ECS service %s, ECS service report may be incomplete: %s", *failure.Arn, *failure.Reason)
+// Returns a channel from which service ARNs can be read.
+// Cannot fail as it will attempt to deliver partial results, though that may end up being no results.
+func (c ecsClientImpl) listServices() <-chan string {
+	log.Debugf("Listing ECS services")
+	results := make(chan string)
+	go func() {
+		count := 0
+		err := c.client.ListServicesPages(
+			&ecs.ListServicesInput{Cluster: &c.cluster},
+			func(page *ecs.ListServicesOutput, lastPage bool) bool {
+				for _, arn := range page.ServiceArns {
+					count++
+					results <- *arn
 				}
+				return true
+			},
+		)
+		if err != nil {
+			log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
+		}
+		log.Debugf("Listed %d services", count)
+		close(results)
+	}()
+	return results
+}
 
-				lock.Lock()
-				for _, service := range resp.Services {
-					results[*service.ServiceName] = service
-				}
-				lock.Unlock()
-			}()
-			return true
-		},
-	)
-	group.Wait()
+// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
+// with full EcsService objects being put into the cache. Closes done when finished.
+func (c ecsClientImpl) describeServices() (chan<- string, <-chan struct{}) {
+	input := make(chan string)
+	done := make(chan struct{})
+
+	log.Debugf("Describing ECS services")
 
+	go func() {
+		const maxServices = 10 // How many services we can put in one Describe command
```

Comment:
Thanks for the clarifications.

>  it was an intentional choice to do each describe in the background so that many describe operations could be executing concurrently (my assumption is that any local processing is much faster than any network call)

Is it worth the compilation? 

Also, is waiting for 10 services really the right thing? 

The channel needs to explicitly closed by the writer in order to "flush" the services so I don't really see the advantage of using a channel for input. You could pass all the services directly (no input channel), chunk them and spawn go routines. Simpler and more explicit.